### PR TITLE
21.0.0.9 GA blog post - Draft

### PR DESCRIPTION
**Pull request by ryan-storey**

### CODE REVIEW

File: posts/2021-09-03-microprofile-21009.adoc [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+:url-about: /
+//Blank line here is necessary before starting the body of the post.
+
+Open Liberty 21.0.0.9 introduces many new benefits such as MicroProfile 4.1 with MicroProfile Health updates. Transaction Peer Recovery allows 2 phase commit to be used in cloud environments for the first time, and a new logging format (`TBASIC`) is made available, as well as a new method to determine active endpoints. This release also comes with many significant bug fixes.
+
+In link:{url-about}[Open Liberty] 21.0.0.9:
+
+* <<mp4_1, MicroProfile 4.1>>
+** <<mpHealth3_1, MicroProfile Health 3.1>>
+* <<tbasic, TBASIC Logging Format Beta>>
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: drop 'Beta'

File: posts/2021-09-03-microprofile-21009.adoc [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+Or take a look at our link:{url-prefix}/downloads/[Downloads page].
+
+[link=https://stackoverflow.com/tags/open-liberty]
+image::img/blog/blog_btn_stack.svg[Ask a question on Stack Overflow, align="center"]
+
+[#mp4_1]
+== MicroProfile 4.1
+
+MicroProfile 4.1 improves developer experience with updates to the Health feature. MicroProfile 4.1 also defines compatible implementation requirements for runtimes, which must pass all of the MicroProfile component specification `TCKs` including Config 2.0, Fault Tolerance 3.0, Rest Client 2.0, Health 3.1, Metrics 3.0, Open Tracing 2.0, Open API 2.0, and JWT propagation 1.2. This beta driver will be used as a compatible implementation for releasing MicroProfile 4.1.
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: ~This beta driver will be used as a compatible implementation for releasing MicroProfile 4.1.~ The previously made available Open Liberty 21.0.0.9-beta release served as the compatible implementation for releasing MicroProfile 4.1, and the 21.0.0.9 release marks our first official release to include the feature.

File: posts/2021-09-03-microprofile-21009.adoc [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+
+To enable the new beta features in your app, add them to your `server.xml`:
+
+[source, xml]
+----
+<feature>mpHealth-3.1</feature>
+----
+
+To find out more viist:
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: visit

File: posts/2021-09-03-microprofile-21009.adoc [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+* link:https://github.com/eclipse/microprofile-health/releases/tag/3.1[MicroProfile Health Check 3.1 Release Page]
+* link:https://github.com/eclipse/microprofile-health/blob/3.1/spec/src/main/asciidoc/release_notes.asciidoc[Release Notes]
+
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc:129]: To get started with and learn more about MicroProfile Health, check out our [MicroProfile Health guide](https://openliberty.io/guides/microprofile-health.html).

File: posts/2021-09-03-microprofile-21009.adoc [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+
+[#tbasic]
+== TBASIC Logging Format
+
+A new logging format, `TBASIC`, is added to match the Websphere Application Server basic logging format. This new format option can be used for `consoleFormat`, `messageFormat`, or `traceFormat`. `TBASIC` format matches the existing `BASIC` format for `traceFormat` (so now either `TBASIC` or `BASIC` can be used to refer to this format for `traceFormat`). Using this new format option will allow users that utilize log parsers that work with the Websphere Application Server basic format to use the same parsers with Open Liberty logs.
+
+The new options can be used in the bootstrap.properties file:
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: wrap file names with backticks, so `bootstrap.properties` file

File: posts/2021-09-03-microprofile-21009.adoc [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+
+[source, java]
+----
+com.ibm.ws.logging.message.format=tbasic
+com.ibm.ws.logging.console.format=tbasic
+com.ibm.ws.logging.trace.format=tbasic
+----
+
+You can also change the format by editing the server.env and adding the following lines:
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: `server.env` file

File: posts/2021-09-03-microprofile-21009.adoc [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+//Blank line here is necessary before starting the body of the post.
+
+Open Liberty 21.0.0.9 introduces many new benefits such as MicroProfile 4.1 with MicroProfile Health updates. Transaction Peer Recovery allows 2 phase commit to be used in cloud environments for the first time, and a new logging format (`TBASIC`) is made available, as well as a new method to determine active endpoints. This release also comes with many significant bug fixes.
+
+In link:{url-about}[Open Liberty] 21.0.0.9:
+
+* <<mp4_1, MicroProfile 4.1>>
+** <<mpHealth3_1, MicroProfile Health 3.1>>
+* <<tbasic, TBASIC Logging Format Beta>>
+* <<method, New method to determine active endpoints>>
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: change 'method' to 'activeEndpoints'

File: posts/2021-09-03-microprofile-21009.adoc [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+[source, java]
+----
+[24/03/21 15:04:10:331 EDT] 00000001 FrameworkMana A   CWWKE0001I: The server defaultServer has been launched.
+[24/03/21 15:04:11:338 EDT] 00000001 FrameworkMana I   CWWKE0002I: The kernel started after 1.177 seconds
+[24/03/21 15:04:11:465 EDT] 0000003e FeatureManage I   CWWKF0007I: Feature update started.
+[24/03/21 15:04:11:635 EDT] 00000033 DropinMonitor A   CWWKZ0058I: Monitoring dropins for applications.
+----
+
+
+[#method]
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: change 'method' to 'activeEndpoints'

File: posts/2021-09-03-microprofile-21009.adoc [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+
+Open Liberty 21.0.0.9 introduces many new benefits such as MicroProfile 4.1 with MicroProfile Health updates. Transaction Peer Recovery allows 2 phase commit to be used in cloud environments for the first time, and a new logging format (`TBASIC`) is made available, as well as a new method to determine active endpoints. This release also comes with many significant bug fixes.
+
+In link:{url-about}[Open Liberty] 21.0.0.9:
+
+* <<mp4_1, MicroProfile 4.1>>
+** <<mpHealth3_1, MicroProfile Health 3.1>>
+* <<tbasic, TBASIC Logging Format Beta>>
+* <<method, New method to determine active endpoints>>
+* <<transaction, Transaction Peer Recovery>>
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc:26]: Make Transaction Peer Recovery the second item in the blog

File: posts/2021-09-03-microprofile-21009.adoc [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+[source, xml]
+----
+<transaction
+...
+recoveryGroup="peer-group-name"
+recoveryIdentity="${HOSTNAME}${wlp.server.name}"
+...
+/>
+----
+
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: For more information regarding the Transaction Manager service, reference the server configuration [docs for transaction](https://openliberty.io/docs/latest/reference/config/transaction.html).

File: posts/2021-09-03-microprofile-21009.adoc [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+As for the fix itself, the `el-3.0` implementation has been updated to select methods as closely as possible to the Java compiler. However, due to ambiguity in the EL spec and the additional EL requirement of type coercion, the EL implementation may select different methods when `varargs` are defined. For more details please read link:https://bz.apache.org/bugzilla/show_bug.cgi?id=65358#c9[comment 9 in the Bugzilla issue] explaining the priority of matching. If an unintended method is selected under these circumstances, we recommend reworking the method definitions.
+The equivalent fix for expressionLangauge-4.0 (currently in beta) will be included in 21.0.0.10. 
+
+== Known issues
+
+In this release, a new issue was discovered. If you would like to find out more, you can check out the corresponding Github issue linked below.
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: link:https://github.com/OpenLiberty/open-liberty/issues/18393[SPNEGO does not work with Java 11 or higher]

In this release, a regression was introduced which prevents the `Krb5LoginModule` class from being found when authenticating using Kerberos on Java 11 or higher. This issue can cause users to be presented with an `Error 403` or `Authentication Failed` when authenticating using the spnego-1.0 feature.

File: posts/2021-09-03-microprofile-21009.adoc [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+recoveryIdentity="${HOSTNAME}${wlp.server.name}"
+...
+/>
+----
+
+
+[#bugs]
+== Notable bugs fixed in this release
+
+We’ve spent some time fixing bugs. The following sections describe just some of the issues resolved in this release. If you’re interested, here’s the  link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A21.0.0.9+label%3A%22release+bug%22[full list of bugs fixed in 21.0.0.9].
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: link is wrong, no periods in version for label:
https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A21009+label%3A%22release+bug%22

**Review by mbroz2 — CHANGES_REQUESTED** [review]

File: posts/2021-09-03-microprofile-21009.adoc (hunk 11 above)
Comment (mbroz2) [reply posts/2021-09-03-microprofile-21009.adoc]: remove the top line "In this release, a new issue was discovered. If you would like to find out more, you can check out the corresponding Github issue linked below."

File: posts/2021-09-03-microprofile-21009.adoc [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+----
+<transaction
+...
+recoveryGroup="peer-group-name"
+recoveryIdentity="${HOSTNAME}${wlp.server.name}"
+...
+/>
+----
+
+For more information regarding the Transaction Manager service, reference the server configuration link:https://openliberty.io/docs/21.0.0.8/reference/config/transaction.html[docs for transaction].
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: in general, all links to docs should use `/latest/` instead of a particular version like `21.0.0.8` 
also, all site links should use {url-prefix} instead of the domain name, 
so you want:
{url-prefix}/docs/latest/reference/config/transaction.html

File: posts/2021-09-03-microprofile-21009.adoc [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+[source, java]
+----
+// Check if the defaultHttpEndpoint is active
+boolean isEndpointActive = mbean.isActive("defaultHttpEndpoint");
+----
+
+For more information about `ServerEndpointControlMBean` methods, refer to the link:https://www.ibm.com/docs/en/was-liberty/nd?topic=SSAW57_liberty/com.ibm.websphere.javadoc.liberty.doc/com.ibm.websphere.appserver.api.kernel.service_1.0-javadoc/com/ibm/websphere/kernel/server/ServerEndpointControlMBean.html[API documentation].
+
+
+[#transaction]
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: Just like you moved this section to the second item in the list at the top, this section needs to be moved to follow the MP section

File: posts/2021-09-03-microprofile-21009.adoc [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+The equivalent fix for expressionLangauge-4.0 (currently in beta) will be included in 21.0.0.10. 
+
+== Known issues
+
+In this release, a new issue was discovered. If you would like to find out more, you can check out the corresponding Github issue linked below.
+
+link:https://github.com/OpenLiberty/open-liberty/issues/18393[SPNEGO does not work with Java 11 or higher]
++
+In this release, a regression was introduced which prevents the Krb5LoginModule class from being found when authenticating using Kerberos on Java 11 or higher. This issue can cause users to be presented with an Error 403 or Authentication Failed when authenticating using the spnego-1.0 feature.
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: `Krb5LoginModule` class 
(aka add backticks)

**Review by mbroz2 — CHANGES_REQUESTED** [review]
couple more minor things, then we can merge into draft

File: posts/2021-09-03-microprofile-21009.adoc [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+
+This issue was fixed by updating the `multipleOf` validation check, to ensure that it is checking that `multipleOf` is greater than zero, rather than one.
+
+* link:https://github.com/OpenLiberty/open-liberty/issues/17682[Exception stack trace is exposed in error returns from JMX REST APIs]
++
+When Liberty's `JMX` REST APIs encounter errors, a JSON object is returned which has a string field labelled 'stackTrace' containing the stack trace from the exception which caused the error. Returning the stack trace was flagged as a security concern, so this field has been removed. An extra string field labelled `error` has been added to the returned `JSON` object, containing the message from the java exception.
+
+* link:https://github.com/OpenLiberty/open-liberty/issues/17912[New Netty 4.1.66 release]
++
+We've pulled in the latest release of the Netty project, which is a dependency of our grpcClient-1.0 feature. This new release includes multiple improvements and bugfixes, for more information see the link:https://netty.io/news/2021/07/16/4-1-66-Final.html[Netty release notes].
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: make "grpcClient-1.0 feature" a link to docs/latest/reference/feature/grpcClient-1.0.html

File: posts/2021-09-03-microprofile-21009.adoc [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+----
+<feature>mpHealth-3.1</feature>
+----
+
+To find out more visit:
+
+* link:https://github.com/eclipse/microprofile-health/releases/tag/3.1[MicroProfile Health Check 3.1 Release Page]
+* link:https://github.com/eclipse/microprofile-health/blob/3.1/spec/src/main/asciidoc/release_notes.asciidoc[Release Notes]
+
+To get started with and learn more about MicroProfile Health, check out our link:https://openliberty.io/guides/microprofile-health.html[MicroProfile Health guide].
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: all site links should use {url-prefix} instead of the domain name, so you want
{url-prefix}/guides/microprofile-health.html

File: posts/2021-09-03-microprofile-21009.adoc [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+For more information regarding the Transaction Manager service, reference the server configuration link:https://openliberty.io/docs/21.0.0.8/reference/config/transaction.html[docs for transaction].
+
+[#bugs]
+== Notable bugs fixed in this release
+
+We’ve spent some time fixing bugs. The following sections describe just some of the issues resolved in this release. If you’re interested, here’s the  link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A21009+label%3A%22release+bug%22[full list of bugs fixed in 21.0.0.9].
+
+* link:https://github.com/OpenLiberty/open-liberty/issues/16700[Improve featureUtility performance with remote repository]
++
+Previously, the `featureUtility` would take a long time to resolve features when using a remote repository with a slow network. Artifacts would be fetched one at a time which slowed things down dramatically. This issue was fixed by fetching and downloading multiple artifacts at the same time from Maven repository.
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: make `featureUtility` also a link to docs/latest/reference/command/command-overview.html#featureUtility

File: posts/2021-09-03-microprofile-21009.adoc [hunk 19]
```diff
@@ -0,0 +1,266 @@
...
+
+* link:https://github.com/OpenLiberty/open-liberty/issues/18237[Ignore FFDC for IOExceptions in handleMessage]
++
+A bug was discovered where first failure data capture (FFDC) files were incorrectly produced. This would happen when bad JSON data was passed to REST APIs that are using the `jaxrs-2.1` feature with `BYO Jackson`. This issue could occur when a JAX-RS resource or provider throws any `IOException`, and was fixed by preventing FFDC from logging the `IOException`.
+
+* link:https://github.com/OpenLiberty/open-liberty/issues/18155[JSF faces-config parser throws NPE when namespace missing]
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc:224]: Previous changes made in order to support Jakarta Faces 3.0 added logic to check for the new `https://jakarta.ee/xml/ns/jakartaee` namespace in `faces-config.xml` files. Because the new logic uses `String.contentEquals()`, if a `faces-config.xml` file declares a version but not a namespace, a `NullPointerException` would be thrown during app startup. This affects all JSF features on Liberty: jsf-2.0, jsf-2.2, jsf-2.3, and faces-3.0.  
To fix this issue, the logic has now been changed to use `String.equals()`. Please note that while the NullPointerException has been fixed, valid `faces-config.xml` descriptors should declare valid namespaces, ie. `http://java.sun.com/xml/ns/javaee`, `http://xmlns.jcp.org/xml/ns/javaee`, or `https://jakarta.ee/xml/ns/jakartaee`.

File: posts/2021-09-03-microprofile-21009.adoc [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+/>
+----
+
+For more information regarding the Transaction Manager service, reference the server configuration link:{url-prefix}/docs/latest/reference/config/transaction.html[docs for transaction].
+
+
+[#tbasic]
+== TBASIC Logging Format
+
+A new logging format, `TBASIC`, is added to match the Websphere Application Server basic logging format. This new format option can be used for `consoleFormat`, `messageFormat`, or `traceFormat`. `TBASIC` format matches the existing `BASIC` format for `traceFormat` (so now either `TBASIC` or `BASIC` can be used to refer to this format for `traceFormat`). Using this new format option will allow users that utilize log parsers that work with the Websphere Application Server basic format to use the same parsers with Open Liberty logs.
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: Change this paragraph to:

A new logging format `TBASIC` has been made available for both `consoleFormat` and `messageFormat`.  Although a `BASIC` logging format already existed for `consoleFormat`, `messageFormat`, and `traceFormat` the output differed between `traceFormat` and `consoleFormat` and `messageFormat`.  The new `TBASIC` format has been added to match the existing `BASIC` format that was already available for `traceFormat`, providing the ability to have consistent formatting across all three logs.  The `TBASIC` format has also been added as a logging format for `traceFormat`, acting as an alias for the existing `BASIC` option.
To summarize, if you want to have the same logging format in `consoleFormat` and `messageFormat` that you could get in `traceFormat` using `BASIC`, you now can by using the `TBASIC` logging format.

**Review by mbroz2 — CHANGES_REQUESTED** [review]

**Review by mbroz2 — DISMISSED** [review]

File: posts/2021-09-03-microprofile-21009.adoc [hunk 21]
```diff
@@ -0,0 +1,268 @@
...
+[source, java]
+----
+WLP_LOGGING_MESSAGE_FORMAT=TBASIC
+WLP_LOGGING_CONSOLE_FORMAT=TBASIC
+----
+
+
+TBASIC Logs Example:
+
+[source, java]
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: this one too should be
[source, role="no_copy"]

File: posts/2021-09-03-microprofile-21009.adoc [hunk 22]
```diff
@@ -0,0 +1,268 @@
...
+
+[#tbasic]
+== TBASIC Logging Format
+
+A new logging format `TBASIC` has been made available for both `consoleFormat` and `messageFormat`.  Although a `BASIC` logging format already existed for `consoleFormat`, `messageFormat`, and `traceFormat` the output differed between `traceFormat` and `consoleFormat` and `messageFormat`.  The new `TBASIC` format has been added to match the existing `BASIC` format that was already available for `traceFormat`, providing the ability to have consistent formatting across all three logs.  The `TBASIC` format has also been added as a logging format for `traceFormat`, acting as an alias for the existing `BASIC` option.
+To summarize, if you want to have the same logging format in `consoleFormat` and `messageFormat` that you could get in `traceFormat` using `BASIC`, you now can by using the `TBASIC` logging format.
+
+The new options can be used in the `bootstrap.properties` file:
+
+[source, java]
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: [source]

File: posts/2021-09-03-microprofile-21009.adoc [hunk 23]
```diff
@@ -0,0 +1,268 @@
...
+----
+com.ibm.ws.logging.message.format=tbasic
+com.ibm.ws.logging.console.format=tbasic
+com.ibm.ws.logging.trace.format=tbasic
+----
+
+You can also change the format by editing the `server.env` file and adding the following lines:
+
+[source, java]
```
Comment (mbroz2) [inline posts/2021-09-03-microprofile-21009.adoc]: [source]

**Review by mbroz2 — CHANGES_REQUESTED** [review]

**Review by mbroz2 — APPROVED** [review]
